Feature/custom tags

xhallix opened the pull request:
@mikermcneil 

this is the very first Idea of https://github.com/balderdashy/sails/issues/3854
I'm going to make smaller PR's to prevent we need to talk about a bunch of code :)
Basically `grunt-sails-linker` is now able to support `type` attribute in the `script` tag.
Further the module itself is responsible for parsing the provided source files.

Example when parsing a node module using `require`:

Gruntfile.js

``` js
'test/fixtures/**/scriptType.html' : require('./test/fixtures/complexScripts').jsFilesToInject
```

ComplexScripts.js

``` js
var jsFilesToInject = [
   {
    src: 'test/fixtures/regularInject/*.js'
  },
  {
    src: [
      'test/fixtures/typeInject/*.js',
      'test/fixtures/typeInject/*.js'
    ],
    type: 'script/babel'
  },
  'js/iammissing/**/*.js'
];

module.exports.jsFilesToInject = jsFilesToInject;
```

As you can see, no more mapping of the module is needed when it will be exported, since `grunt-sails-linker` will take care of that.

**What's next?**

`
 environment: 'production',
    headers: {
      foo: 'bar'
    }
`

I'm looking forward to implement the next nodes, but I'm not 100% sure what you would like to see here.
1. Would you like to modify the way `production` files are loaded and move them out of `tasks/sails-linker.js` ?
2. What about the `headers` node?

Best
xhallix

Commit:
Script type can be set

Commit:
validate through tests

Commit:
Rename testfiles to be more meaningful

Commit:
Rename testfiles to be more meaningful

Commit:
support array script tags

Commit:
Add editor file to gitignore

mikermcneil commented:
@xhallix Meant to look at this sooner-- apologies.  Totally slipped my mind!

Re: mapping, that sounds awesome- will get that stuff out of pipeline.js

Re: what's next:  I think it would be best to have this support arbitrary HTML attributes.  So basically, whatever you put in the dictionary on the RHS gets serialized.  And as for configuration for the socket client via HTML attributes on the `<script>`, sails.io.js should be smart enough to understand JSON automatically.  So in grunt-sails-linker, I think we can do something like:

```js
// ...
// where, in this snippet, **htmlAttrName** is the left hand side (e.g. `'headers'` or `'environment'`, etc -- could be anything!) and **rhs** is the right hand side of it (e.g. `{foo:'bar'}` or `'production'`, could be anything too)...
//


if (!_.isString(rhs)) {
  try {
    rhs = JSON.stringify(rhs);
  } catch (e) {
    throw new Error('Could not stringify the value provided for `'+attrName+'`:'+util.inspect(rhs, {depth:null})+'\nDetails:'+e.stack);
  }
}//>-

//...and then continue on to inject it into the script tag
```

Thoughts?

xhallix commented:
@mikermcneil no problem, think you had a lot on your list with the new sails release 👍 

Yep the abitrary support does make much sense! 
I'm going to improve this, will not be a big deal at all.

I'm doing the same for css then, what do you think?

Regarding the socket client, I'll have a look after this, I'm not very familiar with the config here yet, but I'll keep you updated.

mikermcneil commented:
@xhallix re doing the same for `<link>` tags, sounds good.

>  I'm not very familiar with the config here yet, but I'll keep you updated.

No worries-- it shouldn't matter, since we're approach this as just allowing any HTML attribute (the socket client allows you to set certain config settings that way)

xhallix commented:
Okay great, will go for it within the next week I guess :)

xhallix commented:
@mikermcneil I have this on my list but I'm a Bit busy right now, but I think I will get this done within the next 2 or 3 weeks

mikermcneil commented:
@xhallix Sounds good-- it probably won't make Sails v1.0 in that case, but that's ok (we can add it in a minor version since it's additive).

xhallix commented:
@mikermcneil did you had the chance to take a look at the code yet?
Because for stylesheets I plan to do a `styles-linker` which will do the same for stylesheets as the `script-linker` does. 

Of course I will create a basic `linker` to allow inheritance for the concrete linkers like `script/style -linker`.

Commit:
Grunt-Sails-Linker now supports abitrary html attributes.

Configuration provided in complexScripts.js & complexStyles.js

Extend the test to verify the changes

Commit:
Extend documentation and add FAQ, because using styletags is not obvious

xhallix commented:
@mikermcneil 
support for html attributes and stylesheet support has been added. 
Further I extended the documentation and added needed test.
Please merge if everthing is correct.

Best

mikermcneil commented:
@xhallix thanks!  Taking a look now

sgress454 commented:
Hi @xhallix, sorry this is taking so long.  We still love this feature, but there's a couple of things that are preventing us from merging this in right now:

* We need all code in the Sails codebase to be ES5 for maximum Node version compatibility, so no arrow functions or `const`.  
* Please use `@sailshq/lodash` instead of `lodash`, and make sure it's added to the `dependencies` of `package.json`.   That way it will skip having to install the package in most apps (since it's already a dependency of sails-hook-grunt).
* Is `glob` necessary?  I see it's used in `addArrayScriptTags` and `addStringScriptTags`, but I'm not very clear on what those methods are doing.
* Most importantly, I tried linking this in to an existing Sails project (after installing the missing dependencies) and got:
    ```
    ------------------------------------------------------------------------
    Aborted due to warnings.
    Running "sails-linker:devJs" (sails-linker) task
    Warning: Cannot read property 'replace' of undefined 
    ------------------------------------------------------------------------
    ```
    from Grunt.  I got the same thing with both a Sails v0.12.x and a Sails v1.0 app.  We'll need grunt-sails-linker to continue working with existing apps without users having to make any changes to their  files.

The easiest way to test this with a Sails v0.12.x app is to run `npm link` inside your local `grunt-sails-linker` directory, then:

```
sails new test-sails-linker && cd test-sails-linker
npm link grunt-sails-linker
sails lift --verbose
```

Thanks for all your work on this.  We'll continue to keep an eye on it and merge it in when it's ready!
